types/react/index.d.ts

```diff
@@ -121,8 +122,10 @@ declare namespace React {
         : C extends ((props: any, context?: any) => ReactElement | null)
         ? undefined
         : C extends keyof JSX.IntrinsicElements
-        ? JSX.IntrinsicElements[C] extends DOMAttributes<infer E>
-            ? E
+        ? JSX.IntrinsicElements[C] extends { ref?: infer R }
+            ? R extends RefObject<infer E>
```

Comment:
Ah the previous ones were failing because of the `string` type.

Can you try this instead? It's slightly more robust and covers callback refs as well:

```ts
type ElementRef<C> = C extends keyof JSX.IntrinsicElements 
  ? JSX.IntrinsicElements[C] extends { ref?: infer R } 
    ? Exclude<R, string> extends Ref<infer E> 
      ? E 
      : unknown
    : unknown
  : Rest<C>
```

Reply:
It works! and seems much better. Thanks